File: util/src/main/java/com/greenops/util/datamodel/pipelinestatus/PipelineStatus.java
package com.greenops.util.datamodel.pipelinestatus;

import com.greenops.util.datamodel.auditlog.DeploymentLog;
import com.greenops.util.datamodel.auditlog.Log;

import java.util.ArrayList;
import java.util.List;

public class PipelineStatus {

    private final List<String> progressingSteps;
    private final List<FailedStep> failedSteps;
    private boolean stable;
    private boolean complete;
    private boolean cancelled;

    public PipelineStatus() {
        this.progressingSteps = new ArrayList<>();
        this.stable = true;
        this.complete = true;
        this.cancelled = false;
        this.failedSteps = new ArrayList<>();
    }

    public List<String> getProgressingSteps() {
        return this.progressingSteps;
    }

    public boolean isStable() {
        return this.stable;
    }

    public boolean isComplete() {
        return this.complete;
    }

    public boolean isCancelled() {
        return this.cancelled;
    }

    public List<FailedStep> getFailedSteps() {
        return this.failedSteps;
    }

    public void markCancelled() {
        this.cancelled = true;
        this.complete = false;
    }

    public void markIncomplete() {
        this.complete = false;
    }

    public void addProgressingStep(String stepName) {
        progressingSteps.add(stepName);
    }

    //This method asserts the stability of the pipeline by seeing if the latest audit log shows the state to be stable
    public void addLatestLog(Log log) {
        if (log instanceof DeploymentLog) {
            var deploymentLog = (DeploymentLog) log;

            // deployment failed
            if (deploymentLog.getStatus().equals(Log.LogStatus.FAILURE.name())) {
                this.stable = false;
            }
        } else {
            //Was instance of a RemediationLog
            if (!log.getStatus().equals(Log.LogStatus.FAILURE.name())) {
                this.stable = false;
            }
            //TODO: Needs to be added to failed step
        }
    }

    public void addLatestDeploymentLog(DeploymentLog log, String step) {
        if (log.getStatus().equals(Log.LogStatus.FAILURE.name())) {
            this.complete = false;
            addFailedDeploymentLog(log, step);
        }
    }

    public void addFailedDeploymentLog(DeploymentLog log, String step) {
        this.failedSteps.add(new FailedStep(step, !log.isDeploymentComplete(), log.getBrokenTest(), log.getBrokenTestLog()));
    }
}

File: workflowtrigger/pipelinestatus/pipeline_status.go
package pipelinestatus

import (
	"github.com/greenopsinc/util/auditlog"
	"gitlab.com/c0b/go-ordered-json"
)

// TODO: should be replaced with PipelineSpec from CRD
type PipelineStatus struct {
	ProgressingSteps []string     `json:"progressingSteps"`
	Stable           bool         `json:"stable"`
	Complete         bool         `json:"complete"`
	Cancelled        bool         `json:"cancelled"`
	FailedSteps      []FailedStep `json:"failedSteps"`
}

func New() PipelineStatus {
	return PipelineStatus{
		ProgressingSteps: make([]string, 0),
		Stable:           true,
		Complete:         true,
		Cancelled:        false,
		FailedSteps:      make([]FailedStep, 0),
	}
}

func (p *PipelineStatus) MarkCancelled() {
	p.Cancelled = true
	p.Complete = false
}

func (p *PipelineStatus) MarkIncomplete() {
	p.Complete = false
}

func (p *PipelineStatus) AddProgressingStep(stepStatus string) {
	p.ProgressingSteps = append(p.ProgressingSteps, stepStatus)
}

func (p *PipelineStatus) AddLatestLog(log auditlog.Log) {
	switch log.(type) {
	case *auditlog.DeploymentLog:
		deploymentLog := log.(*auditlog.DeploymentLog)
		if deploymentLog.GetStatus() == auditlog.Failure {
			p.Stable = false
		}
	case *auditlog.RemediationLog:
		if log.(*auditlog.RemediationLog).GetStatus() != auditlog.Failure {
			p.Stable = false
		}
	}
}

func (p *PipelineStatus) AddLatestDeploymentLog(log auditlog.DeploymentLog, step string) {
	if log.GetStatus() == auditlog.Failure {
		p.Complete = false
		p.AddFailedDeploymentLog(log, step)
	}
}

func (p *PipelineStatus) AddFailedDeploymentLog(log auditlog.DeploymentLog, step string) {
	p.FailedSteps = append(p.FailedSteps, FailedStep{
		Step:             step,
		DeploymentFailed: !log.DeploymentComplete,
		BrokenTest:       log.BrokenTest,
		BrokenTestLog:    log.BrokenTestLog,
	})
}

func MarshallPipelineStatus(status PipelineStatus) *ordered.OrderedMap {
	mapObj := ordered.NewOrderedMap()

	var failedStepsList []*ordered.OrderedMap
	failedStepsList = make([]*ordered.OrderedMap, 0)
	for _, val := range status.FailedSteps {
		step := MarshallFailedStep(val)
		failedStepsList = append(failedStepsList, step)
	}

	mapObj.Set("progressingSteps", status.ProgressingSteps)
	mapObj.Set("stable", status.Stable)
	mapObj.Set("complete", status.Complete)
	mapObj.Set("cancelled", status.Cancelled)
	mapObj.Set("failedSteps", failedStepsList)
	return mapObj
}
